include/nccl_ofi_freelist.h

```diff
@@ -38,6 +51,7 @@ struct nccl_ofi_freelist_block_t {
 	void *memory;
 	size_t memory_size;
 	void *mr_handle;
+	nccl_ofi_freelist_elem_t *entries;
```

Comment:
I don't think we need this?

Reply:
Before, we stored the entry metadata inside the user data. This patch is attempting to separate the two, which means we need to allocate the entries separately.

An alternative would be to store the entries with the freelist, instead of with the block, and expand the freelist's entry store when we add new blocks. I don't have a strong opinion either way.

Reply:
It doesn't appear you ever use this field, and just jam the entries directly into the freelist.

Reply:
> It doesn't appear you ever use this field, and just jam the entries directly into the freelist.

Ah, right, we do put the entries directly in the freelist. The reason we need the `block->entries` pointer is to free the entries on finalize (see `nccl_ofi_freelist.c:190` in this PR). We allocate the entries in groups (`nccl_ofi_freelist.c:272`) rather than individually, so just having the pointers to the entries themselves (`nccl_ofi_freelist_t::entries`) isn't enough information to know which pointers to free.

One thing we could do here is have `block->entries` be a flexible array member of `nccl_ofi_freelist_block_t`, so we remove the extra degree of pointer chasing. However, I'm not sure C++ has any valid equivalent of flexible array members...

Reply:
i don't understand?  you're just freeing the blocks during cleanup.

Reply:
> i don't understand? you're just freeing the blocks during cleanup.

I also free `block->entries` before freeing the block. https://github.com/aws/aws-ofi-nccl/pull/600/files#diff-9f90753ff38be494f380c1245a96b589aa782c96d212b3cdf536baa44df9fa06R190